doc: Use CONFIG_SITE variable instead of --prefix option

hebasto opened the pull request:
The current examples of `--prefix=...` option usage to point `configure` script to appropriate `depends` directory is not [standard](https://www.gnu.org/prep/standards/html_node/Directory-Variables.html). This causes some [confusion](https://github.com/bitcoin/bitcoin/pull/16691) and a bit of inconvenience.

Consider a CentOS 7 32 bit system. Packages `libdb4-devel`, `libdb4-cxx-devel`, `miniupnpc-devel` and `zeromq-devel` are unavailable from repos. After recommended build with depends:
```
cd depends
make
cd ..
./autogen.sh
./configure --prefix=$PWD/depends/i686-pc-linux-gnu
make
```
a user is unable to `make install` compiled binaries neither locally (to `~/.local`) nor system-wide (to `/usr/local`) as `--prefix` is set already.

Meanwhile, the standard approach with using [`config.site`](https://www.gnu.org/software/automake/manual/html_node/config_002esite.html) files allows both possibilities:

```
cd depends
make
cd ..
./autogen.sh
CONFIG_SITE=$PWD/depends/i686-pc-linux-gnu/share/config.site ./configure --prefix ~/.local
make
make install
```

or 

```
CONFIG_SITE=$PWD/depends/i686-pc-linux-gnu/share/config.site ./configure
make
sudo make install  # install to /usr/local
```

Moreover, this approach is used in [Gitian descriptors](https://github.com/bitcoin/bitcoin/tree/master/contrib/gitian-descriptors) already.

hebasto commented:
@practicalswift 
https://github.com/bitcoin/bitcoin/issues/14959#issuecomment-487413672:
> FWIW you're not alone here: I found the use of `--prefix` (for anything beyond the ordinary meaning of installation destination) unintuitive too :-)

https://github.com/bitcoin/bitcoin/pull/16691#issuecomment-524170556:
> Should it be clarified that `--prefix` is used not in its usual meaning of "when you run `./configure --prefix=/foo/bar`, you are indicating that the software should be installed under the `/foo/bar` directory"?
>
> That seems to be a recurring source of confusion.
>
> Has it been discussed if we could use something else than `--prefix` for the "where to find dependencies"-meaning we're talking about here?

ping @ryanofsky @fanquake

ryanofsky commented:
This change seems clearer to me and I think it's an improvement. But now the documentation doesn't seem to mention the prefix path at all, so it's not clear that binaries will be installed in `depends/<host-platform-triplet>`. It might be worth mentioning in these documents where depends build outputs will be installed if the prefix is no longer explicit in the instructions

hebasto commented:
> This change seems clearer to me and I think it's an improvement. But now the documentation doesn't seem to mention the prefix path at all, so it's not clear that binaries will be installed in `depends/<host-platform-triplet>`. It might be worth mentioning in these documents where depends build outputs will be installed if the prefix is no longer explicit in the instructions

Do you mean copy-paste from `./configure --help` ?

ryanofsky commented:
> Do you mean copy-paste from `./configure --help` ?

I'm suggesting adding a sentence like: "The default install prefix when using `config.site` is `--prefix=depends/<host-platform-triplet>`, so depends build outputs will be installed in that location."

Without this PR, the install location is more obvious because --prefix is explicit. The install location is less clear with this PR because it stops mentioning --prefix.

fanquake commented:
~0. It has annoyed me for a while that we've been inconsistent with `--prefix` and `CONFIG_SITE` usage in our docs, and I was going to change the windows docs to use `--prefix` at one point.

> Consider a CentOS 7 32 bit system.

Is this the only system where this is an issue?

I know @theuni is busy, but I'd like his thoughts here as well.

hebasto commented:
@fanquake 

> ~0. It has annoyed me for a while that we've been inconsistent with `--prefix` and `CONFIG_SITE` usage in our docs, and I was going to change the windows docs to use `--prefix` at one point.

Using `--prefix` in a such [unusual](https://www.gnu.org/prep/standards/html_node/Directory-Variables.html) way looks a bit like a hack. And the windows docs is ok, IMO.

> > Consider a CentOS 7 32 bit system.
> 
> Is this the only system where this is an issue?

This issue is not CentOS specific as using `--prefix` make it impossible to `make install` compiled binaries neither locally (to ~/.local) nor system-wide (to /usr/local) on every system.
But on other systems a user could get binaries in other ways (https://bitcoincore.org/en/download/).

hebasto commented:
@ryanofsky 
> I'm suggesting adding a sentence like: "The default install prefix when using `config.site` is `--prefix=depends/<host-platform-triplet>`, so depends build outputs will be installed in that location."

IIUC, "depends build outputs" (boost, qt, zmq, ...) _always_ are placed in `depends/<host-platform-triplet>` directory.
And `prefix` is used by `make install` as a target to copy binaries and libraries from `src` tree.

What are reasons to install to `depends/<host-platform-triplet>`?
`/home/hebasto/github/bitcoin/depends/x86_64-pc-linux-gnu/bin/bitcoind` seems not very useful.
`/home/hebasto/.local/bin/bitcoind` or `/usr/local/bitcoind` looks much better, IMO.

> Without this PR, the install location is more obvious because --prefix is explicit. The install location is less clear with this PR because it stops mentioning --prefix.

Without this PR mentioning `prefix` was mandatory as its usage was non-standard and there was a discrepancy between our docs and autotools docs (and `configure --help`).

With this PR a user who going to use `make install` command could use our depends docs, and `prefix` in a usual and well-documented (by autotools) way to point to a desired install location.

ryanofsky commented:
Code review ACK c8e20035b088afeb731f4427e2f874e3291388a8. Agree that referencing the config.site file specifically is a not only clearer than current prefix magic but also more flexible, as in in the centos example.

I don't understand this comment at all: https://github.com/bitcoin/bitcoin/pull/17934#issuecomment-575015186. I like the PR. I'm an only trying to suggest adding another sentence to the documentation. But it's fine not to add it.

hebasto commented:
@ryanofsky Sorry, Russel. I did not want to be rude. My native language is not English, as you know. I was confused with "depends build outputs will be installed in that location" phrase. Actually, I do not understand its meaning now.

hebasto commented:
> Agree that referencing the config.site file specifically is a not only clearer than current prefix magic but also more flexible, as in in the centos example.

More details from [The Autoconf Manual - 15.8](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.69/html_node/Site-Defaults.html):
> If the environment variable `CONFIG_SITE` is set, `configure` uses its value as the name of a shell script to read; it is recommended that this be an absolute file name. Otherwise, it reads the shell script _prefix_`/share/config.site` if it exists, then _prefix_`/etc/config.site` if it exists.

Sjors commented:
Concept ACK.

Unfortunately `CONFIG_SITE` isn't remembered between configure runs. This causes a problem when you switch branches or pull master sometimes, when configure is automatically re-run, it forgets about using depends.

hebasto commented:
@Sjors 
> Unfortunately `CONFIG_SITE` isn't remembered between configure runs. This causes a problem when you switch branches or pull master sometimes, when configure is automatically re-run, it forgets about using depends.

Do you mean that `--prefix=...` option is remembered in mentioned circumstances?

Sjors commented:
I believe so yes. Though hopefully there's a way to make it remember the config_site too.

hebasto commented:
@theuni from #8167:
> To make my life easier while I was in there, I fixed up depends to play nice with CONFIG_SITE, which can be used now instead of abusing --prefix, though the --prefix hack still works fine.

fanquake commented on depends/README.md:
```diff
@@ -12,15 +12,16 @@ For example:
 order for it to pick up libraries, tools, and settings from the depends build,
-you must point it at the appropriate `--prefix` directory generated by the
-build. In the above example, a prefix dir named x86_64-w64-mingw32 will be
+you must set `CONFIG_SITE` environment variable to appropriate
```
you must set the `CONFIG_SITE` environment variable to point to a `config.site` settings file. In the above example, a file named `depends/x86_64-w64-mingw32/share/config.site` will be created. To use it during compilation:

CONFIG_SITE=$PWD/depends/x86_64-w64-mingw32/share/config.site ./configure

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/17934#issuecomment-816556534).

fanquake reviewed (commented):
Can you rebase this, and update the use of `--prefix` in the multiprocess documentation.

hebasto committed:
doc: Use CONFIG_SITE instead of --prefix

hebasto commented:
Updated c8e20035b088afeb731f4427e2f874e3291388a8 -> 223b1ba7d90509a47ea07af46f4b9c3b8efbc9f8 ([pr17934.01](https://github.com/hebasto/bitcoin/commits/pr17934.01) -> [pr17934.02](https://github.com/hebasto/bitcoin/commits/pr17934.02)):

- rebased
- addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/17934#issuecomment-574869188):
> I'm suggesting adding a sentence like: "The default install prefix when using `config.site` is `--prefix=depends/<host-platform-triplet>`, so depends build outputs will be installed in that location."

- addressed the recent @fanquake's comments

practicalswift commented:
Concept ACK

Thanks for fixing this. I've always found our use of `--prefix` somewhat unintuitive since it deviates from the typical meaning.

practicalswift commented:
ACK 223b1ba7d90509a47ea07af46f4b9c3b8efbc9f8: patch looks correct

fanquake commented on depends/README.md:
```diff
@@ -12,15 +12,18 @@ For example:
 
-Common `host-platform-triplets` for cross compilation are:
+The default install prefix when using `config.site` is `--prefix=depends/<host-platform-triplet>`,
+so depends build outputs will be installed in that location.
```
By "depends build outputs" I'm assuming you mean things like the fontconfig tools, which will end up installed into `depends/<host-platform-triplet>`. i.e:
```bash
# ls depends/x86_64-pc-linux-gnu/bin/
event_rpcgen.py  fc-cache  fc-cat  fc-list  fc-match  fc-pattern  fc-query  fc-scan  fc-validate  freetype-config  xmlwf
```

Is that correct?

hebasto replied:
All of the dependency packages are installed into `depends/<host-platform-triplet>`, using `bin`, `lib` and other subdirectories.

The `make -C depends` rightfully reports:
```
copying packages: native_b2 boost libevent qt expat libxcb xcb_proto libXau xproto freetype fontconfig libxkbcommon qrencode bdb sqlite miniupnpc libnatpmp zeromq
to: /home/hebasto/GitHub/bitcoin/depends/x86_64-pc-linux-gnu
```

---

Not to be confused with the default `--prefix` for Bitcoin Core, that is `/usr/local`. That means `make install` will install `bitcoind` into `/usr/local/bin`.

fanquake reviewed (approved):
ACK 223b1ba7d90509a47ea07af46f4b9c3b8efbc9f8

hebasto commented:
@Sjors 
> Unfortunately `CONFIG_SITE` isn't remembered between configure runs. This causes a problem when you switch branches or pull master sometimes, when configure is automatically re-run, it forgets about using depends.

Passing `CONFIG_SITE` to `configure` as a parameter
```
$ ./configure --prefix ~/.local CONFIG_SITE=$PWD/depends/i686-pc-linux-gnu/share/config.site
```
should work in such cases.
